### ISSUE TITLE
False positive for duplicated code

### ISSUE THREAD

**Issue by fcoedno** [golangci/golangci-lint#1372]
Thank you for creating the issue!

- [ X ] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [ X ] Yes, I've searched similar issues on GitHub and didn't find any.
- [ X ] Yes, I've included all information below (version, config, etc).

Please include the following information:

<details><summary>Version of golangci-lint</summary>

```console
$ golangci-lint --version
1.31.0
```

</details>

<details><summary>Config file</summary>

```console
$ cat .golangci.yml
# Full list on https://golangci-lint.run/usage/linters/
linters:
  disable-all: true
  enable:
    #- exhaustive    # check exhaustiveness of enum switch statements
    #- funlen        # Tool for detection of long functions
    #- gocognit      # Computes and checks the cognitive complexity of functions
    #- godot         # Check if comments end in a period
    #- golint        # Golint differs from gofmt. Gofmt reformats Go source code, whereas golint prints out style mistakes
    #- lll           # Reports long lines
    #- scopelint     # Scopelint checks for unpinned variables in go programs
    - bodyclose     # checks whether HTTP response body is closed successfully
    - deadcode      # Finds unused code
    - depguard      # Go linter that checks if package imports are in a list of acceptable packages
    - dupl          # Tool for code clone detection
    - errcheck      # Errcheck is a program for checking for unchecked errors in go programs. These unchecked errors can be critical bugs in some cases
    - exportloopref # checks for pointers to enclosing loop variables
    - goconst       # Finds repeated strings that could be replaced by a constant
    - gocritic      # The most opinionated Go source code linter
    - gocyclo       # Computes and checks the cyclomatic complexity of functions
    - gofmt         # Gofmt checks whether code was gofmt-ed. By default this tool runs with -s option to check for code simplification
    - goimports     # Goimports does everything that gofmt does. Additionally it checks unused imports
    - gosec         # Inspects source code for security problems
    - gosimple      # Linter for Go source code that specializes in simplifying a code
    - govet         # Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string
    - ineffassign   # Detects when assignments to existing variables are not used
    - interfacer    # Linter that suggests narrower interface types
    - maligned      # Tool to detect Go structs that would take less memory if their fields were sorted
    - misspell      # Finds commonly misspelled English words in comments
    - nakedret      # Finds naked returns in functions greater than a specified function length
    - noctx         # noctx finds sending http request without context.Context
    - prealloc      # Finds slice declarations that could potentially be preallocated
    - rowserrcheck  # checks whether Err of rows is checked successfully
    - staticcheck   # Staticcheck is a go vet on steroids, applying a ton of static analysis checks
    - structcheck   # Finds unused struct fields
    - stylecheck    # Stylecheck is a replacement for golint
    - typecheck     # Like the front-end of a Go compiler, parses and type-checks Go code
    - unconvert     # Remove unnecessary type conversions
    - unparam       # Reports unused function parameters
    - unused        # Checks Go code for unused constants, variables, functions and types
    - varcheck      # Finds unused global variables and constants
    - whitespace    # Tool for detection of leading and trailing whitespace
```

</details>

<details><summary>Go environment</summary>

```console
$ go version && go env
go version go1.15.1 linux/amd64
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/edno/.cache/go-build"
GOENV="/home/edno/.config/go/env"
GOEXE=""
GOFLAGS=" -mod=vendor"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY="bitbucket.org/arquivei/*"
GONOSUMDB="bitbucket.org/arquivei/*"
GOOS="linux"
GOPATH="/home/edno/go"
GOPRIVATE="bitbucket.org/arquivei/*"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/edno/.asdf/installs/golang/1.13/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/edno/.asdf/installs/golang/1.13/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/edno/Projects/foundationkit/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build017524104=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details><summary>Verbose output of running</summary>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /app / /root] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 31 linters: [bodyclose deadcode depguard dupl errcheck exportloopref goconst gocritic gocyclo gofmt goimports gosec gosimple govet ineffassign interfacer maligned misspell nakedret noctx prealloc rowserrcheck staticcheck structcheck stylecheck typecheck unconvert unparam unused varcheck whitespace] 
INFO [loader] Go packages loading at mode 575 (exports_file|files|name|compiled_files|deps|imports|types_sizes) took 45.840469359s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 8.571039ms 
INFO [linters context/goanalysis] analyzers took 1m7.522930972s with top 10 stages: buildir: 41.49542648s, buildssa: 3.489710964s, inspect: 1.792445671s, dupl: 1.636421711s, fact_deprecated: 1.5070619s, printf: 1.331778585s, ctrlflow: 1.012949528s, unconvert: 975.517967ms, fact_purity: 866.841747ms, interfacer: 854.457177ms 
INFO [linters context/goanalysis] analyzers took 2.154322585s with top 10 stages: buildir: 1.983550239s, U1000: 170.772346ms 
INFO [runner] Issues before processing: 173, after processing: 3 
INFO [runner] Processors filtering stat (out/in): exclude-rules: 51/51, sort_results: 3/3, exclude: 51/173, diff: 3/3, max_per_file_from_linter: 3/3, max_same_issues: 3/3, skip_files: 173/173, identifier_marker: 173/173, skip_dirs: 173/173, max_from_linter: 3/3, source_code: 3/3, path_shortener: 3/3, severity-rules: 3/3, cgo: 173/173, filename_unadjuster: 173/173, nolint: 3/51, uniq_by_line: 3/3, path_prefixer: 3/3, path_prettifier: 173/173, autogenerated_exclude: 173/173 
INFO [runner] processing took 20.431167ms with stages: exclude: 9.165065ms, identifier_marker: 4.541283ms, nolint: 2.669445ms, autogenerated_exclude: 1.734767ms, path_prettifier: 1.437612ms, skip_dirs: 707.158µs, source_code: 102.149µs, cgo: 33.922µs, filename_unadjuster: 20.408µs, max_same_issues: 5.372µs, max_from_linter: 4.05µs, uniq_by_line: 3.838µs, path_shortener: 2.448µs, max_per_file_from_linter: 1.105µs, skip_files: 590ns, diff: 517ns, exclude-rules: 500ns, sort_results: 359ns, severity-rules: 315ns, path_prefixer: 264ns 
INFO [runner] linters took 12.517840814s with stages: goanalysis_metalinter: 11.761792896s, unused: 735.274909ms 
sefaz/accesskey/validation.go:10: 10-46 lines are duplicate of `sefaz/accesskey/validation.go:49-83` (dupl)
func Check(accessKey AccessKey) error {
        const op errors.Op = "accesskey.Check"

        if accessKey == "" {
                return errors.E(op, ErrEmptyAccessKey, ErrCodeEmptyAccessKey)
        }

        if len(accessKey) != 44 {
                return errors.E(op, ErrInvalidLength, ErrCodeInvalidLength)
        }

        if !isDigitOnly(accessKey) {
                return errors.E(op, ErrInvalidCharacter, ErrCodeInvalidCharacter)
        }

        if !isValidUF(accessKey[0:2].String()) {
                return errors.E(op, ErrInvalidUF, ErrCodeInvalidUF)
        }

        if !isValidMonth(accessKey[4:6].String()) {
                return errors.E(op, ErrInvalidMonth, ErrCodeInvalidMonth)
        }

        if !isValidCPFCNPJ(accessKey[6:20].String()) {
                return errors.E(op, ErrInvalidCPFCNPJ, ErrCodeInvalidCPFCNPJ)
        }

        if !isValidModel(accessKey[20:22].String()) {
                return errors.E(op, ErrInvalidModel, ErrCodeInvalidModel)
        }

        if !isValidationDigitCorrect(accessKey.String()) {
                return errors.E(op, ErrInvalidDigit, ErrCodeInvalidDigit)
        }

        return nil
}
sefaz/accesskey/validation.go:49: 49-83 lines are duplicate of `sefaz/accesskey/validation.go:10-46` (dupl)
func validate(accessKey AccessKey) error {
        const op errors.Op = "validate"
        if accessKey == "" {
                return errors.E(op, ErrEmptyAccessKey, ErrCodeEmptyAccessKey)
        }
        if len(accessKey) != 44 {
                return errors.E(op, ErrInvalidLength, ErrCodeInvalidAccessKey)
        }

        if !isDigitOnly(accessKey) {
                return errors.E(op, ErrInvalidCharacter, ErrCodeInvalidAccessKey)
        }

        if !isValidUF(accessKey[0:2].String()) {
                return errors.E(op, ErrInvalidUF, ErrCodeInvalidAccessKey)
        }

        if !isValidMonth(accessKey[4:6].String()) {
                return errors.E(op, ErrInvalidMonth, ErrCodeInvalidAccessKey)
        }

        if !isValidCPFCNPJ(accessKey[6:20].String()) {
                return errors.E(op, ErrInvalidCPFCNPJ, ErrCodeInvalidAccessKey)
        }

        if !isValidModel(accessKey[20:22].String()) {
                return errors.E(op, ErrInvalidModel, ErrCodeInvalidAccessKey)
        }

        if !isValidationDigitCorrect(accessKey.String()) {
                return errors.E(op, ErrInvalidDigit, ErrCodeInvalidAccessKey)
        }

        return nil
}
sefaz/stakeholder/stakeholder.go:22:1: commentFormatting: put a space between `//` and comment text (gocritic)
//UnmarshalJSON deserializes a JSON value into a Stakeholder value
^
INFO File cache stats: 106 entries of total size 205.2KiB 
INFO Memory: 585 samples, avg is 140.4MB, max is 545.5MB 
INFO Execution took 58.378276798s                 
```

</details>

Although the reported piece of duplicated code shares the very same algorithm structure, at a closer look to the output we can see that the return values of each if branch is different, which makes it a false-positive for code duplication.

**Comment by ldez** [comment]
Hello,

I think it's not a bug but the expected behavior, dupl use a specific method to find duplicates.

From [dupl documentation](https://github.com/mibk/dupl#dupl-):
> It just operates with their types (e.g. `if a == 13 {}` and `if x == 100 {}` are
considered the same provided it exceeds the minimal token sequence size).
>
>Due to the used method dupl can report so called "false positives" on the output. These are
the ones we do not consider clones (whether they are too small, or the values of the matched
tokens are completely different).

So you have to change the threshold.

example:
```yml
linters-settings:
  dupl:
    # tokens count to trigger issue, 150 by default
    threshold: 200
```

**Comment by Udhayarajan** [comment]
with linter v2

```yml
linters:
  default: none
  enable:
    - dupl
  settings:
    dupl:
      threshold: 200
```
